dev/Interop/CameraCaptureUI/CameraCaptureUI/CameraCaptureUI.cpp

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+#include "pch.h"
+#include <future>
+#include <ppltasks.h>
+#include <pplawait.h>
+#include <ShlObj.h>
+#include "CameraCaptureUI.h"
+#include "Microsoft.Windows.Media.Capture.CameraCaptureUI.g.cpp"
+#include "CameraCaptureUITelemetry.h"
+#include "TelemetryHelper.h"
+
+//using namespace winrt::Windows::Media::Capture;
+using namespace winrt::Windows::UI;
+using namespace winrt::Windows::Storage;
+using namespace winrt::Windows::System;
+using namespace winrt::Windows::Foundation;
+using namespace winrt::Windows::Foundation::Collections;
+using namespace winrt::Windows::ApplicationModel::DataTransfer;
+
+namespace winrt::Microsoft::Windows::Media::Capture::implementation
+{
+    IAsyncOperation<hstring> LaunchCameraForResultToken(winrt::Microsoft::UI::WindowId const& targetWindow, ValueSet const& properties)
+    {
+        LauncherOptions options;
+        if (targetWindow.Value)
+        {
+            check_hresult(options.as<IInitializeWithWindow>()->Initialize(reinterpret_cast<HWND>(targetWindow.Value)));
+        }
+
+        options.TreatAsUntrusted(false);
+        options.DisplayApplicationPicker(false);
+        options.TargetApplicationPackageFamilyName(L"Microsoft.WindowsCamera_8wekyb3d8bbwe");
+
+        Uri launchUri{ L"microsoft.windows.camera.picker:" };
+        winrt::Windows::System::LaunchUriResult result = co_await Launcher::LaunchUriForResultsAsync(launchUri, options, properties);
+        if (!result.Result())
+        {
+            co_return L"";
+        }
+        else if (auto t = result.Result().TryLookup(L"SelectedTokens"))
+        {
+            co_return t.as<hstring>();
+        }
+        else
+        {
+            throw hresult_canceled();
+        }
+    }
+
+    concurrency::task<token_and_path> CreateEmptyFileAndGetToken(hstring const& fileExtension)
+    {
+        auto tempFolder = ApplicationData::Current().TemporaryFolder();
+        auto tempFile = co_await tempFolder.CreateFileAsync(L"CCapture" + fileExtension, CreationCollisionOption::GenerateUniqueName);
+        auto tempFilePath = tempFile.Path();
+        auto token = SharedStorageAccessManager::AddFile(tempFile);
+
+        co_return token_and_path{ token, tempFilePath };
+    }
+
+    IAsyncOperation<StorageFile> CameraCaptureUI::CaptureFileAsync(CameraCaptureUIMode mode)
+    {
+        bool isAppPackaged = m_telemetryHelper.IsPackagedApp();
+        PCWSTR appName = m_telemetryHelper.GetAppName().c_str();
+        try
+        {
```

Comment:
There should be a lock to prevent/hold parallel calls to CaptureFileAsync

Reply:
If two (or more) consecutive CaptureFileAsync async calls run well, no change required.

Interested to know if this is a real use case for the developers since existing UWP API had issues (with lock) itself but not aware if there is bug raised for the same. Great to know if such feedback comes from the preview program. Can be taken up once the preview feedback is received.

Reply:
The problem with current implementation without lock is that parallel calls to CaptureFileAsync API accidently overwrites the values of m_photoTokenFile and m_videoTokenFile member variables depending upon the mode of capture. That seems fundamentally wrong as the first call is not finished and tokens are already overwritten, and UWP API prevent that by lock (of-course that lock is not working as per expectation, needs to be fixed separately).
Either we can address this by modifying the API implementation and make it safe for parallel calls or write a lock and restrict it.

Reply:
If values are overridden, below piece of code will be executed and the result will be as expected.

`file = co_await SharedStorageAccessManager::RedeemTokenForFileAsync(tokenValue);`

Your concern is valid. However, as per the conversation with Saharsh, the API gives expected result so probably a follow up task or waiting to hear from preview feedback is ok.

If the below statement works well, why we need to have `m_photoTokenFile` and `m_videoTokenFile` member variables. Can we remove these two variable?

`file = co_await SharedStorageAccessManager::RedeemTokenForFileAsync(tokenValue);`

Reply:
I understand that the Camera app has certain limitations, and the team recognized this when deciding to implement a lock to ensure only one capture operation occurs at a time on the original API. Even if multiple capture operations are initiated, only the focused camera will render, while others will show a message stating "Camera is already in use by another app."
Also, for real world use cases, if people do actually want multiple camera capture instances up, they should go through a feature request process.
For now, I am going ahead with the Locking mechanism and restricting this behavior just as it's done for Original API.

Reply:
Another idea was discussed is can `m_photoTokenFile` and `m_videoTokenFile`  be converted to local variable to avoid lock?

Reply:
In addition to the camera app's limitations, the main concern is that the token can get overwritten if locking mechanism is not there. While suggesting that we make the token variables local could eliminate the need for a lock, this approach also presents risks. Without a lock, modifying properties during an ongoing operation—such as changing the extension type and calling CaptureFileAsync again—can lead to unpredictable behavior. The outcome often depends on which call reaches the camera app first, resulting in randomness that we should avoid.
 
Furthermore, testing revealed that while two parallel calls to the camera app may work fine, initiating three or more calls leads to a faulty state, ultimately requiring the app to be force-closed. This highlights a potential bug in Camera app's implementation, but it also shows why we need a locking mechanism.
 
To ensure stable and predictable user experiences, I will proceed with adding the lock to prevent multiple capture operations from happening at the same time.